model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/AbstractExtendableRevisioned.java

```diff
@@ -16,14 +16,36 @@
  */
 package org.keycloak.models.cache.infinispan.entities;
 
+import org.jboss.logging.Logger;
+
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
 
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public abstract class AbstractExtendableRevisioned extends AbstractRevisioned {
-    protected ConcurrentHashMap cachedWith = new ConcurrentHashMap();
+
+    private static final int UNINITIALISED = 0;
+    private static final int INITIALISING = 1;
+    private static final int INITIALISED = 2;
+
+    private static Logger logger = Logger.getLogger(AbstractExtendableRevisioned.class);
+
+    private static ThreadLocal<Boolean> weInitTL = ThreadLocal.withInitial(() -> false);
+
+    protected CoordinatingMap cachedWith = new CoordinatingMap(this);
```

Comment:
Due to lack of time, I'd rather have the changes in this sensitive part as conservative as possible. Hence while I believe introducing the `CoordinatingMap` might be good long-term, we might want to omit this now since that adds too much new untested complexity. For 3.4.2 to get ready soon, the original approach seemed to fix the most likely causes.

(I'd even rather only add the `AtomicBoolean initialized` to the AbstractExtendableRevisioned and have the ReentrantLock be allocated dynamically by some per-user lock provider, but that optimization I'd leave for 4.x.)

The `CoordinatingMap` also increases memory footprint compared to the `ConcurrentHashMap` which, together with three new properties, can negatively impact the performance since the number of users might be very large so we should try to keep the object as light memory-wise as possible.

Reply:
`CoordinatingMap` could indeed impact performance, hard to tell without doing some performance metrics. Its current implementation uses inheritance, but it would be better to use delegation approach so ensureInited() will be guaranteed to only be called once for every method - I'll change the implementation.

About `ReentrantLock` I don't see how you can optimise it with lazy init (if I understand correctly) since that would in itself require an extra locking object as you need multiple threads to share the same instance of the lock.
The benefit of `CoordinatingMap` is that it should (properly done) fix the problem completely -  no need for CredentialProvider to call `ensureInited()` before accessing `cachedWith`. But yes, for now we can find all CredentialProviders and fix their code. That won't fix race condition for any third party custom providers.

@hmlnarik If I understand correctly you prefer for now my original PR proposal with synchronised blocks and without CoordinatingMap, and suggest to have a CoordinatingMap + ReentrantLock implementation in some branch and working on performance testing it?

Reply:
@mstruk Almost - I'd prefer the original proposal with **synchronized** replaced by `ReentrantLock`, and discuss any further optimizations after release. Why should `CredentialProvider` need to call `ensureInited()` at all? That all should be part of the intitialization in `UserCacheSession` - and if that does not succeed in given time, adding it to individual methods of the `cachedWith` does not help (rather the opposite, it might wait several times the given time, since there might be several calls to the `cachedWith` methods, and the lock can be held for quite a time).

Reply:
Stian and I had discussion with Marco over BJ.  We think that calling oncache directly after creating a CachedUser and before attempting to add this instance to the cache is the simpler approach and should probably fix the problem....I hope...

Reply:
+1 That would be the safest - provided there is no additional caching or some other global state change in bodies of onCache methods (we can control it on our side and we seem to be safe here but may represent an issue in custom extensions). We should probably add a note to "onCache" javadocs that the method can be called concurrently for different instances of the same user but only one of these instances would eventually end up in the cache.
Thinking aloud, should we rename it from onCache to e.g. onBeforeCache and add a notification invoked after the object is actually put into the cache e.g. onCached? I personally don't see much value in it, and it might be probably done in 4.x as well if needed.

Reply:
It could have had a better name to make it clearer, but we can't just rename it as this is a supported SPI. We should rather handle it with making sure the JavaDocs are clear.